### PR TITLE
feat: add huber loss from rlax.

### LINKED ISSUE #729: [FEATURE] Huber loss
### Please describe the purpose of the feature. Is it related to a problem?
Implement huber loss for the JAX IPPO system.

### PR DESCRIPTION
**Pull request by siddarthsingh1**
## What?
Huber loss added to IPPO.
## Why?
Further optimisation of MAVA's IPPO algorithm
## How?
Add option to use a huber loss instead of MSE.
## Extra
Closed #729 
[Benchmarking results](https://www.notion.so/instadeep/Huber-loss-for-value-function-b3b0ae1d1c21423a8da8b3b589bc6065)

### CODE REVIEW

**Comment by DriesSmit** [comment]
Thanks, @siddarthsingh1 :slightly_smiling_face: Can you just link the issue as well :+1:

File: mava/systems/jax/ippo/system.py [hunk 1]
```diff
@@ -94,3 +94,71 @@ def design(self) -> Tuple[DesignSpec, Any]:
             component_dependency_guardrails=ComponentDependencyGuardrails,
         )
         return system, default_params
+
```
Comment (OmaymaMahjoub) [inline mava/systems/jax/ippo/system.py:97]: Change the name of the file to systems or separate the two systems each in a py file since we have a full folder called systems

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Looks good thanks @siddarthsingh1, it will be interesting if you add test case for the huber loss in the unit tests of `lesses_test.py` (whether in this PR or you can make a new PR for it)

File: tests/jax/components/training/losses_test.py [hunk 2]
```diff
@@ -239,7 +239,6 @@ def test_mapg_loss_separate_networks(
     low_loss_critic = low_agent_0_critic_loss["loss_critic"]
 
     assert jnp.isclose(loss_entropy, -0.47500002)
-    assert loss_critic == 4.5
```
Comment (DriesSmit) [inline tests/jax/components/training/losses_test.py:242]: Can you maybe add a test for the Huber loss? Instead of needing to remove this assert?

File: mava/components/jax/training/losses.py [hunk 3]
```diff
@@ -166,6 +168,252 @@ def policy_loss_fn(
                 )
             return policy_grads, loss_info_policy
 
+        def critic_loss_grad_fn(
```
Comment (DriesSmit) [inline mava/components/jax/training/losses.py:171]: Just maybe chat to @KaleabTessera about this. I think he was thinking one could provide a function to the system instead of having to duplicate the loss code?

File: examples/jax/debugging/simple_spread/feedforward/decentralised/run_ippo_huber.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    policy_optimiser = optax.chain(
+        optax.clip_by_global_norm(40.0), optax.scale_by_adam(), optax.scale(-1e-4)
+    )
+
+    critic_optimiser = optax.chain(
+        optax.clip_by_global_norm(40.0), optax.scale_by_adam(), optax.scale(-1e-4)
+    )
+
+    # Create the system.
+    system = ippo.IPPOSystemHuber()
```
Comment (DriesSmit) [inline examples/jax/debugging/simple_spread/feedforward/decentralised/run_ippo_huber.py:94]: Maybe instead of creating a new system we can just do something like [this](https://github.com/instadeepai/Mava/blob/feature/recurrent-actors/examples/jax/debugging/simple_spread/recurrent/decentralised/run_ippo.py#L95). Otherwise we might need a new system for every configuration of the components.

File: mava/systems/jax/ippo/system.py [hunk 5]
```diff
@@ -94,3 +94,71 @@ def design(self) -> Tuple[DesignSpec, Any]:
...
+
+class IPPOSystemHuber(System):
```
Comment (DriesSmit) [inline mava/systems/jax/ippo/system.py:99]: We can probably remove this system and just add the Huber component in the example.

**Review by DriesSmit — CHANGES_REQUESTED** [review]
Thanks @siddarthsingh1 :fire: Nice work :slightly_smiling_face:

File: examples/jax/debugging/simple_spread/feedforward/decentralised/run_ippo_huber.py (hunk 4 above)
Comment (siddarthsingh1) [reply examples/jax/debugging/simple_spread/feedforward/decentralised/run_ippo_huber.py]: Updated it to add the huber loss through the update instead of a separate system.

File: mava/systems/jax/ippo/system.py (hunk 5 above)
Comment (siddarthsingh1) [reply mava/systems/jax/ippo/system.py]: Removed.

File: tests/jax/components/training/losses_test.py (hunk 2 above)
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:232]: I've added tests for the separated functions and readded the assert.

File: mava/systems/jax/ippo/system.py (hunk 1 above)
Comment (siddarthsingh1) [reply mava/systems/jax/ippo/system.py]: The extra system has been removed

File: mava/components/jax/training/losses.py (hunk 3 above)
Comment (siddarthsingh1) [reply mava/components/jax/training/losses.py]: I'll ask about that. I might have misinterpreted what was being discussed :sweat_smile: .

File: tests/jax/components/training/losses_test.py (hunk 2 above)
Comment (DriesSmit) [reply tests/jax/components/training/losses_test.py:232]: Thanks :)

File: examples/jax/debugging/simple_spread/feedforward/decentralised/run_ippo_huber.py (hunk 4 above)
Comment (DriesSmit) [reply examples/jax/debugging/simple_spread/feedforward/decentralised/run_ippo_huber.py]: Thanks :slightly_smiling_face:

File: mava/components/jax/training/losses.py (hunk 3 above)
Comment (DriesSmit) [reply mava/components/jax/training/losses.py]: Thanks :)

**Review by DriesSmit — DISMISSED** [review]
Thanks @OmaymaMahjoub :fire:

**Comment by OmaymaMahjoub** [comment]
> Thanks @OmaymaMahjoub fire

Thanks @DriesSmit for helping with the redesign!

File: mava/systems/config.py [hunk 6]
```diff
@@ -136,13 +137,13 @@ def update(self, **kwargs: Any) -> None:
                         self._config[name] = dataclass
                 else:
                     raise Exception(
-                        f"The given component config ({name}) is not part of the current \
```
Comment (DriesSmit) [inline mava/systems/config.py]: Why was this removed?

File: mava/systems/config.py [hunk 7]
```diff
@@ -85,7 +85,9 @@ def add(self, **kwargs: Any) -> None:
                             self._param_to_component[new_param_name] = name
                     self._config[name] = dataclass
             elif isinstance(dataclass, SimpleNamespace):
-                self._config[name] = dataclass
+                # SimpleNamespace implies that
```
Comment (DriesSmit) [inline mava/systems/config.py:88]: Why was this changed?

File: mava/systems/system.py [hunk 8]
```diff
@@ -58,22 +57,17 @@ def _make_config(self) -> None:
...
-            self.config.add(**input)
 
-    @abc.abstractmethod
-    def design(self) -> Tuple[DesignSpec, Dict]:
-        """System design specifying the list of components to use.
-
-        Returns:
-            Tuple[system callback components, system config].
-        """
+            if config_class is not SimpleNamespace:
```
Comment (DriesSmit) [inline mava/systems/system.py:61]: Why did we add an if statement here? We still need to config right?

Comment (DriesSmit) [inline mava/systems/system.py:61]: Why did we add an if statement here? We still need to config right?

File: mava/systems/config.py (hunk 6 above)
Comment (OmaymaMahjoub) [reply mava/systems/config.py]: It was because merge conflict edit I will restore it

File: mava/systems/system.py (hunk 8 above)
Comment (OmaymaMahjoub) [reply mava/systems/system.py:61]: I just passed the systempy from develop branch

Comment (OmaymaMahjoub) [reply mava/systems/system.py:61]: I will check how to fix it

[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply mava/systems/system.py]: This related to the develop branch it is not a change made in this PR :/

File: mava/components/component_groups.py [hunk 9]
```diff
@@ -1,7 +1,12 @@
...
 
 # The components that are needed for a recurrent policy.
 recurrent_policy_components = [
     executing.RecurrentExecutorSelectAction,
     executing.RecurrentExecutorObserve,
 ]
+
+# The components that a needed for IPPO with the Huber value loss.
+huber_value_loss_components = [
+    training.HuberValueLossFunction,
```
Comment (arnupretorius) [inline mava/components/component_groups.py:11]: If there is just one component for the loss, should we be adding it to a "groups" file? 

Also, I think the group components here are general (since they are located inside the general components folder), so we can remove the string saying huber is just for IPPO, since other systems might want to use it as well.

File: mava/components/training/losses.py [hunk 10]
```diff
@@ -28,6 +29,91 @@
 from mava.core_jax import SystemTrainer
 
 
+class DefaultValueLossFunction(Component):
```
Comment (arnupretorius) [inline mava/components/training/losses.py:32]: Since this is setting up a SE loss, I think we should rename this to `SquaredErrorLoss`? We can also drop the "Function" in the class names to make them shorter.

File: mava/components/training/losses.py [hunk 11]
```diff
@@ -28,6 +29,91 @@
...
+            training.BaseTrainerInit
+        ]  # import from training to avoid partial dependency
+
+
+@dataclass
+class HuberValueLossFunctionConfig:
+    huber_delta: float = 1.0
+
+
+class HuberValueLossFunction(DefaultValueLossFunction):
```
Comment (arnupretorius) [inline mava/components/training/losses.py:86]: Let's drop "Function" in class name. Also, not sure why we need to inherit from the default/SE loss? I think what we should do is create an abstract loss, as done below with the `Loss` class, but then be more specific with the difference between these classes. Might be good to have a quick offline chat about this? 🙂

**Review by arnupretorius — CHANGES_REQUESTED** [review]
Thanks for this! 👍 Made a few suggestions 🙂

File: mava/components/training/losses.py (hunk 10 above)
Comment (KaleabTessera) [reply mava/components/training/losses.py:32]: +1

File: tests/components/training/losses_test.py [hunk 12]
```diff
@@ -275,3 +288,99 @@ def test_mapg_loss(
...
     assert loss_critic == 4.5
     assert low_loss_critic < loss_critic
+
+
+def test_mapg_huber_loss(
+    mock_trainer: Trainer,
+    mapg_loss: MAPGWithTrustRegionClippingLoss,  # noqa: E501
+) -> None:
+    """Test whether mapg huber loss output is as expected"""
+    huber_loss = HuberValueLossFunction()
```
Comment (KaleabTessera) [inline tests/components/training/losses_test.py:298]: Can we set `huber_delta` in the test? Incase the default ever changes (we still want this test to pass).

File: mava/components/training/losses.py (hunk 11 above)
Comment (OmaymaMahjoub) [reply mava/components/training/losses.py:86]: I think this could be a great option too, @DriesSmit  what do you think?

File: tests/components/training/losses_test.py (hunk 12 above)
Comment (OmaymaMahjoub) [reply tests/components/training/losses_test.py:298]: Yeah sure!

**Review by KaleabTessera — COMMENTED** [review]
Thanks @siddarthsingh1 @DriesSmit and @OmaymaMahjoub ! This is very close to being ready from my perspective!

File: tests/components/training/losses_test.py (hunk 12 above)
Comment (OmaymaMahjoub) [reply tests/components/training/losses_test.py:302]: I made a test to check that it update the value in the config: https://github.com/instadeepai/Mava/blob/451b5f896b07610eee46c7487dd743d5c81f25b0/tests/components/training/losses_test.py#L175

File: mava/components/training/losses.py (hunk 11 above)
Comment (DriesSmit) [reply mava/components/training/losses.py]: Sounds good to me :slightly_smiling_face:

File: mava/components/component_groups.py (hunk 9 above)
Comment (DriesSmit) [reply mava/components/component_groups.py:11]: Agreed. I removed the IPPO text. I added the Huber component to a component group to standardise our component setup. But I don't have a strong opinion on whether it should be there or not. I'm happy to remove it if you think that is better :+1:

File: tests/components/training/losses_test.py (hunk 12 above)
Comment (KaleabTessera) [reply tests/components/training/losses_test.py]: Thanks!

**Review by KaleabTessera — DISMISSED** [review]

File: mava/components/component_groups.py (hunk 9 above)
Comment (DriesSmit) [reply mava/components/component_groups.py]: I removed the component_group as we discussed :)

File: mava/components/training/losses.py [hunk 13]
```diff
@@ -29,12 +29,28 @@
...
+    def name() -> str:
+        """_summary_
+
+        Returns:
+            _description_
+        """
+        return "value_loss"
+
+
+class SquaredErrorLoss(ValueLoss):
```
Comment (arnupretorius) [inline mava/components/training/losses.py:48]: Sorry, I know I suggest the original name change, and this is nitpicky, but should we change it to `SquaredErrorValueLoss` to be more consistent?

File: mava/components/component_groups.py [hunk 14]
```diff
@@ -6,7 +6,7 @@
     executing.RecurrentExecutorObserve,
 ]
 
-# The components that a needed for IPPO with the Huber value loss.
+# The components that a needed for the Huber value loss.
 huber_value_loss_components = [
```
Comment (arnupretorius) [inline mava/components/component_groups.py]: Spoke to @DriesSmit about this offline: that we should remove single components from the groups file. This is just a reminder to do that before we merge.

**Review by arnupretorius — COMMENTED** [review]
Last few comments 😉

File: mava/components/component_groups.py (hunk 14 above)
Comment (arnupretorius) [reply mava/components/component_groups.py]: Ok this is done :)

File: mava/components/training/losses.py (hunk 13 above)
Comment (arnupretorius) [reply mava/components/training/losses.py:48]: Or change `HuberValueLoss` to `HuberLoss`

Comment (OmaymaMahjoub) [reply mava/components/training/losses.py:48]: Okay I will change the Square error to `SquaredErrorValueLoss`

Comment (OmaymaMahjoub) [reply mava/components/training/losses.py:48]: just we need to wait for the benchmark to end

Comment (OmaymaMahjoub) [reply mava/components/training/losses.py]: Done!

**Review by KaleabTessera — APPROVED** [review]
Thanks everyone!

**Review by arnupretorius — APPROVED** [review]
Thanks! 🔥
